lib/spack/spack/environment.py

```diff
@@ -498,6 +500,45 @@ def regenerate(self, all_specs, roots):
         view.add_specs(*add_specs, with_dependencies=False)
 
 
+def _strategy(func):
```

Comment:
@alalazo 

I'm curious why you take this approach to ensuring that a function is available to process the desired strategy.  To be clear, this looks correct but I think it's more of a maintenance burden than just maintaining an "if" statement inside of `Environment.concretize`. The schema validator will ensure that the user enters an allowed concretization scheme. If you add an exception at the end like

```
if method == 'together':
  ...
elif method == 'separately':
  ...
else:
  raise Exception("Unrecognized concretization scheme: {0}".format(method)) # Raise something more specific than Exception
```

That should also handle the case where the user adds a method to the schema but forgets to handle it with a new method.

So I think this should be simplified but let me know if you see issues with that. Otherwise LGTM

Reply:
> So I think this should be simplified but let me know if you see issues with that. Otherwise LGTM

Before  the implementation was dynamic, consistent by design and less code but with a slightly more cryptic schema :slightly_smiling_face:  All in all I would vote for the dynamic approach, but I'm really fine with any of the three approaches proposed / implemented during the review. Let me know if you want me to change the code like you propose above.

A few considerations:

1. Discussing with @becker33 the if clause might become longer (like 4 branches at least?) making it not really pretty in my view to treat it with an `if`/`else` statement. 

2. With the dynamic code schema and Environment class were consistent by design. Right now we check that `Environment` has all the implementations it needs at module load time - an inconsistent state will be surely caught by CI as it will make Spack not function at all. With the approach you propose we'll get an exception raised during concretization, which means we could leave inconsistent state if we don't stress the code with a unit test.

3. Using a decorator helps point maintainers to the fact that certain methods implement concretization strategies. It's a more visible marker than just a docstring.

Reply:
In short, I would prefer either an if/else or a dictionary like

```
{'together': function1, 'separately': function2}
```

in more detail:

> Discussing with becker33 the if clause might become longer (like 4 branches at least?) making it not really pretty in my view to treat it with an if/else statement.

IMO an if statement with 4 branches is fine if it handles 4 different things, but I also think a simpler approach would be to have a dictionary from strategy to function like:

```
concretize_fn = stategies.get(method, None)
if not concretize_fn:
  raise Exception("No such strategy:...")
```

> With the approach you propose we'll get an exception raised during concretization, which means we could leave inconsistent state if we don't stress the code with a unit test.

I think it's worth it to make it simpler and I think unit tests ought to be covering these cases anyhow (which they currently do I think). Ultimately for something more obscure/dynamic than this I could see the benefit of this approach but IMO for something that will change infrequently like this the blunt approach has the most appeal to me.

> Using a decorator helps point maintainers to the fact that certain methods implement concretization strategies. It's a more visible marker than just a docstring.

IMO the if/else statement or dictionary would also document this sufficiently

Reply:
I'll go with if/else. With the dictionary I need to modify `concretize` anyhow to add a new method. Anyhow, nitpicks :wink: